05-structs.Rmd

```diff
@@ -0,0 +1,467 @@
+    File Mutect_AnnotatedTable = annovar.output_annotated_table
+  }
+}
+
+
+
+
+
+
+# TASK DEFINITIONS
+
+# Align fastq file to the reference genome
+task BwaMem {
+  input {
+    File input_fastq
+    referenceGenome refGenome
+  }
+  
+  String base_file_name = basename(input_fastq, ".fastq")
+  String ref_fasta_local = basename(refGenome.ref_fasta)
+
+  String read_group_id = "ID:" + base_file_name
+  String sample_name = "SM:" + base_file_name
+  String platform = "illumina"
+  String platform_info = "PL:" + platform   # Create the platform information
+
+
+  command <<<
+    set -eo pipefail
+
+    #can we iterate through a struct??
+    mv ~{refGenome.ref_fasta} .
+    mv ~{refGenome.ref_fasta_index} .
+    mv ~{refGenome.ref_dict} .
+    mv ~{refGenome.ref_amb} .
+    mv ~{refGenome.ref_ann} .
+    mv ~{refGenome.ref_bwt} .
+    mv ~{refGenome.ref_pac} .
+    mv ~{refGenome.ref_sa} .
+
+    bwa mem \
+      -p -v 3 -t 16 -M -R '@RG\t~{read_group_id}\t~{sample_name}\t~{platform_info}' \
+      ~{ref_fasta_local} ~{input_fastq} > ~{base_file_name}.sam 
+    samtools view -1bS -@ 15 -o ~{base_file_name}.aligned.bam ~{base_file_name}.sam
+    samtools sort -@ 15 -o ~{base_file_name}.sorted_query_aligned.bam ~{base_file_name}.aligned.bam
+  >>>
+
+  output {
+    File analysisReadySorted = "~{base_file_name}.sorted_query_aligned.bam"
+  }
+  
+  runtime {
+    memory: "48 GB"
+    cpu: 16
+    docker: "fredhutch/bwa:0.7.17"
+  }
+}
+
+# Mark duplicates (not SPARK, for some reason that does something weird)
+task MarkDuplicates {
+  input {
+    File input_bam
+  }
+
+  String base_file_name = basename(input_bam, ".sorted_query_aligned.bam")
+  String output_bam = "~{base_file_name}.duplicates_marked.bam"
+  String output_bai = "~{base_file_name}.duplicates_marked.bai"
+  String metrics_file = "~{base_file_name}.duplicate_metrics"
+
+  command <<<
+    gatk MarkDuplicates \
+      --INPUT ~{input_bam} \
+      --OUTPUT ~{output_bam} \
+      --METRICS_FILE ~{metrics_file} \
+      --CREATE_INDEX true \
+      --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 \
+      --VALIDATION_STRINGENCY SILENT
+  >>>
+
+  runtime {
+    docker: "broadinstitute/gatk:4.1.4.0"
+    memory: "48 GB"
+    cpu: 4
+  }
+
+  output {
+    File markDuplicates_bam = "~{output_bam}"
+    File markDuplicates_bai = "~{output_bai}"
+    File duplicate_metrics = "~{metrics_file}"
+  }
+}
+
+# Base quality recalibration
+task ApplyBaseRecalibrator {
+  input {
+    File input_bam
+    File input_bam_index
+    File dbSNP_vcf
+    File dbSNP_vcf_index
+    File known_indels_sites_VCFs
+    File known_indels_sites_indices
+    referenceGenome refGenome
+  }
+  
+  String base_file_name = basename(input_bam, ".duplicates_marked.bam")
+  
+  String ref_fasta_local = basename(refGenome.ref_fasta)
+  String dbSNP_vcf_local = basename(dbSNP_vcf)
+  String known_indels_sites_VCFs_local = basename(known_indels_sites_VCFs)
+
+
+  command <<<
+  set -eo pipefail
+
+  mv ~{refGenome.ref_fasta} .
```

Comment:
we should be able to make a bash array with all of these files, including dbSNP_vcf, known_indels_sites_VCFs, etc.